service/src/main/java/com/epam/esm/service/GiftCertificateService.java

```diff
@@ -0,0 +1,180 @@
+        }
+        return list;
+    }
+    /**
+     * Send request for getting all Certificates which contain substring in name or description
+     *
+     * @param substr - String
+     * @return List of Dto which contains Certificates which contain substring *
+     */
+    public List<GiftAndTagDto> getCertificatesBySubstr(String substr) {
+        List<GiftAndTagDto> list = giftCertificateDao.findByNameOrDescriptionContaining(substr);
+        if (list.isEmpty()) {
+            throw new ResourceNotFoundException(String.format("Gift Certificates with substring \'%s\' into name or description not found", substr));
+        }
+        return list;
+    }
+    /**
+     * Send request for getting all sorted Certificates with their tags
+     *
+     * @param sort - String. Type of sort.
+     * @return List of sorted Dto *
+     */
+    public List<GiftAndTagDto> sortCertificates(String sort) {
+        List<GiftAndTagDto> list = giftCertificateDao.findAll();
+        if (list.isEmpty()) {
+            throw new ResourceNotFoundException("Gift Certificates not found");
+        }
+        switch (sort) {
+            case "name-ASC":
+                return list.stream()
+                        .sorted(Comparator.comparing(GiftAndTagDto::getName))
+                        .collect(Collectors.toList());
+            case "name-DESC":
+                return list.stream()
+                        .sorted(Comparator.comparing(GiftAndTagDto::getName)
+                                .reversed())
+                        .collect(Collectors.toList());
+            case "date-ASC":
+                return list.stream()
+                        .sorted(Comparator.comparing(GiftAndTagDto::getCreate_date))
+                        .collect(Collectors.toList());
+            case "date-DESC":
+                return list.stream()
+                        .sorted(Comparator.comparing(GiftAndTagDto::getCreate_date)
+                                .reversed())
+                        .collect(Collectors.toList());
+            case "ASC":
+                return list.stream()
+                        .sorted(Comparator.comparing(GiftAndTagDto::getName)
+                                .thenComparing(GiftAndTagDto::getCreate_date))
+                        .collect(Collectors.toList());
+            case "DESC":
+                return list.stream()
+                        .sorted(Comparator.comparing(GiftAndTagDto::getName)
+                                .thenComparing(GiftAndTagDto::getCreate_date)
+                                .reversed())
+                        .collect(Collectors.toList());
+            default:
+                return list;
+        }
+    }
+    /**
+     * Send request for saving GiftAndTagDto
+     *
+     * @param giftAndTagDto - instance of GiftAndTagDto
+     */
+    public void save(GiftAndTagDto giftAndTagDto) {
```

Comment:
Transactional better to be placed on Service layer, as service operation may fail in the end, but data will be already committed